### ISSUE TITLE
event gap error on ship restart

### ISSUE THREAD

**Issue by baudtack** [urbit/urbit#1406]
My ship crashed today. I still have the pier hanging around. if that would be helpful to inspect.
```
urb@david:~$ urbit rapfyr-diglyt
~
urbit 0.8.0
boot: home is /home/urb/rapfyr-diglyt
loom: mapped 2048MB
jesus my lite: arvo formula 6f2c0b2f
lite: core 3b567bcf
lite: final state 3b567bcf
loom: mapped 2048MB
boot: protected loom
live: loaded: MB/335.134.720
boot: installed 263 jets
work: play 239736
---------------- playback starting----------------
pier: load: commit: event gap: 3a878, b
disk error: pier: load: comit: event gap
lmdb: aborting replay due to error.
Failed to read event log for replay. Exiting...pier: exit
```
Any thoughts on how to recover this would be great.

**Comment by baudtack** [comment]
Should have included this to start with. This is when the original crash happened.

~rapfyr-diglyt> |bonk
~rapfyr-diglyt> fml
  |
>
[~ u=~2019.7.30..22.00.21..c9aa]
~[[date=~2019.7.30..22.00.21..c9aa duct=[i=/a/ames t=~[//newt/0v1u.i8k04]]]]
> |bonk
>=
~palfun-foslup> pretty sure autobonk exists ~rapfyr, making |bonk obsolete
~rapfyr-diglyt> well shit's all fuckedup  so
~rapfyr-diglyt:talk> trying urb@david:~$

**Comment by joemfb** [comment]
@baudtack, you've apparently made it past this issue. Do you have anything else to add to this report?

FWIW, I believe the underlying issue is that this error can be spurious if an event happens to be enqueued too early. We need more granular state transitions when initializing the pier.

**Comment by baudtack** [comment]
I came back and someone... had fix it? or it fixed it self by... magic? i honestly have no idea what the hell happened but yeah it works now. Should have closed this since it's not really a thing anymore. 
I'll reopen if it happens again. sorry for the confusion.

**Comment by philipcmonk** [comment]
I saw this once on a fake ship.   Trying again seemed to work fine.

**Comment by eglaysher** [comment]
My current working theory on this is that it fails if you added events too early. I've encountered this whenever I've started typing during restart.

**Comment by belisarius222** [comment]
I hit something like this just now on a fakezod.  "event gap" error on restart after Ctrl-Z, followed by successful startup on second try.

https://pastebin.com/d8dJjjr9

**Comment by belisarius222** [comment]
I was typing during restart, which would support @eglaysher's theory.

**Comment by belisarius222** [comment]
I'm reopening this.  Just because the ship starts again doesn't make this issue solved.  I'll rename the title to be more specific now that we know more about it.

**Comment by joemfb** [comment]
It's really spurious to check this condition in `pier.c` as written -- it's possible that I/O drivers are enabled before all events-to-be replayed are checked. This falls under the more general issue that boot/startup needs to go through more explicit state transitions. When and how to fix this has to be weighed against other priorities and other ongoing work ...

The ordering of events is already checked in assertions in the worker process. I think this check in pier.c should be removed.

**Comment by joemfb** [comment]
This is fixed in #3064.
